xapian-applications/omega/handler_libextractor.cc

```diff
@@ -0,0 +1,136 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <extractor.h>
+#include <cstring>
+
+using namespace std;
+
+struct metadata {
+    string& title;
+    string& author;
+    string& keywords;
+    string& pages;
+
+    metadata(string& t, string& a, string& k, string& p)
+	    : title(t), author(a), keywords(k), pages(p) {}
+};
+
+/*
+* cls: passed as last parameter from EXTRACTOR_extract
+* plugin_name: name of the plugin
+* type: mime-type of file according to libextractor
+* format: format information about data
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
+static int
+process_metadata(void* cls,
+		 const char* plugin_name,
+		 enum EXTRACTOR_MetaType type,
+		 enum EXTRACTOR_MetaFormat format,
+		 const char* data_mime_type,
+		 const char* data,
+		 size_t data_len)
+{
+    struct metadata* md = (struct metadata*)cls;
+
+    switch (format) {
+	case EXTRACTOR_METAFORMAT_UTF8:
+	    break;
+
+	default:
+	    // specific encoding unknown
+	    // EXTRACTOR_METAFORMAT_UNKNOWN
+	    // EXTRACTOR_METAFORMAT_BINARY
+	    // EXTRACTOR_METAFORMAT_C_STRING
+	    return 0;
+    }
+
+    switch (type) {
+	case EXTRACTOR_METATYPE_BOOK_TITLE:
+	case EXTRACTOR_METATYPE_JOURNAL_NAME:
+	case EXTRACTOR_METATYPE_ORIGINAL_TITLE:
+	case EXTRACTOR_METATYPE_SUBJECT:
+	case EXTRACTOR_METATYPE_TITLE:
+	    if (!md->title.empty())
+		md->title.append(" ");
+	    md->title.append(data, data_len);
+	    break;
+
+	case EXTRACTOR_METATYPE_PAGE_COUNT:
+	    md->pages.assign(data, data_len);
+	    break;
+
+	case EXTRACTOR_METATYPE_ARTIST:
+	case EXTRACTOR_METATYPE_AUTHOR_NAME:
+	case EXTRACTOR_METATYPE_CREATOR:
+	case EXTRACTOR_METATYPE_MOVIE_DIRECTOR:
+	case EXTRACTOR_METATYPE_ORIGINAL_ARTIST:
+	case EXTRACTOR_METATYPE_ORIGINAL_WRITER:
+	case EXTRACTOR_METATYPE_PUBLISHER:
+	    if (!md->author.empty())
+		md->author.append(" ");
+	    md->author.append(data, data_len);
+	    break;
+
+	default:
+	    if (!md->keywords.empty())
+		md->keywords.append(" ");
+	    md->keywords.append(data, data_len);
+    }
+    return 0;
+}
+
+bool
+extract(const string& filename,
+	const string& mimetype,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	struct metadata md(title, author, keywords, pages);
+
+	// Add all default plugins
+	struct EXTRACTOR_PluginList* plugins
+	 = EXTRACTOR_plugin_add_defaults(EXTRACTOR_OPTION_DEFAULT_POLICY);
```

Comment:
Xapian coding style is to wrap after `=` not before.

But you can probably avoid needing to wrap by using `auto` for the type.  The exact type isn't interesting here - this is really just an opaque thing representing plugins which we need to pass to other libextractor API functions.

Reply:
I have changed the type to `auto`. But I had to still wrap the lines as it was still exceeding 80 chars.